### PR TITLE
Workaround for misaligned SegmentTimeline refreshes

### PR DESCRIPTION
**Pull request by glennguy**
Addresses #703, since we know now there is at least one live manifest packager out there that produces incorrect manifests and a backend that serves up segments with arbitrary timestamps.

The most informative comment in that thread is https://github.com/xbmc/inputstream.adaptive/issues/703#issuecomment-860048289 . It shows how subsequent manifest refreshes are providing timestamps that are slightly out of alignment to where they should be. As the times are slightly ahead the current code in IA will increment one time too many and the new segments are mapped incorrectly against the old.

The way this 'fix' works is that if our segment timestamp (provided by manifest refresh) does not exactly line up with an existing segment timestamp then as long as we're very close (within 2% compared to next segment length) to the existing timestamp we assume that was the intended one.

As far as I can tell from DASH-IF documents SegmentTimeline is meant to be *exact*, not an estimate and the previous code acted on that assertion. New code allows a small amount of flexibility and allows something to work that would otherwise be broken, while not affecting properly constructed streams.

### CODE REVIEW

**Comment by ratcashdev** [comment]
THIS WORKS!

**Comment by ratcashdev** [comment]
@glennguy  Is this part of the latest release (.19)?

**Comment by glennguy** [comment]
No, it needs to be merged first before it will become part of a release.  I just need to update the OP slightly then I'll ask for a review. Once accepted then it can be merged and made part of a release. Not long to go now :)

**Comment by glennguy** [comment]
@matthuisman could you take a look and review for me please?

**Review by matthuisman — COMMENTED** [review]
Looks good!

**Review by matthuisman — APPROVED** [review]

File: src/common/AdaptiveStream.h [hunk 1]
```diff
@@ -90,6 +90,8 @@ namespace adaptive
     bool write_data(const void *buffer, size_t buffer_size);
     bool prepareDownload(const AdaptiveTree::Segment *seg);
     adaptive::AdaptiveTree& GetTree() { return tree_; };
+    virtual void SetLastUpdated(std::chrono::system_clock::time_point tm){};
```
Comment (phunkyfish) [inline src/common/AdaptiveStream.h]: Space before braces: `tm) {};`

**Review by phunkyfish — APPROVED** [review]
Just the one, very minor, minor.

**Pull request merged**